t/admin/ssl4.t

```diff
@@ -55,8 +56,23 @@ location /t {
 
             local sess, err = sock:sslhandshake(nil, "www.test.com", true)
             if not sess then
-                ngx.say("failed to do SSL handshake: ", err)
-                return
+                sock = ngx.socket.tcp()
```

Comment:
Why do we need a retry for FIPS?
BTW, we can exact the connect operation into a function, instead of repeating it.

Reply:
In openssl 3.0, decrypt error would be saved in global, which is different to openssl 1.1.
The test case provides two decrypt keys, the first is wrong one, it test the retries and success finally.

https://github.com/apache/apisix/blob/master/t/admin/ssl4.t#L238

```
=== TEST 6: client request successfully, use the two key_encrypt_salt to decrypt in turn
--- yaml_config
apisix:
    node_listen: 1984
    ssl:
        key_encrypt_salt:
            - edd1c9f0985e76a2
            - edd1c9f0985e76a1
```

Then there comes an awkward problem.

The cosocket tries to do ssl hanshake:

https://github.dev/apache/apisix/blob/master/t/admin/ssl4.t#L56

```
            local sess, err = sock:sslhandshake(nil, "www.test.com", true)
            if not sess then
                ngx.say("failed to do SSL handshake: ", err)
                return
            end
```

In the server side of the same worker process, it failed to decrypt using the first key, which would save the error in the `SSL` object.

https://github.com/apache/apisix/blob/master/apisix/ssl.lua#L173

```lua
local function parse_pem_priv_key(sni, pkey)
    core.log.debug("parsing priv key for sni: ", sni)

    local parsed, err = ngx_ssl.parse_pem_priv_key(aes_decrypt_pkey(pkey))
    return parsed, err
end


function _M.fetch_pkey(sni, pkey)
    local parsed_pkey, err = pkey_cache(pkey, nil, parse_pem_priv_key, sni, pkey)
    if not parsed_pkey then
        return false, err
    end

    return parsed_pkey
end
```

Then when the client side handshake returns, it would get error, but in fact, the handshake is successful, it jsut gets the decrypt error saved before.

`src/event/ngx_event_openssl.c`

```c
n = SSL_do_handshake(c->ssl->connection);
...
sslerr = SSL_get_error(c->ssl->connection, n);
```

When you do second handshake, because the server side reused the decrypted key in the cache, so no problem then.

All in all, it's triggered by the combination of cosocket and server reside in the same worker process, openssl changes and nginx ssl handshake handling.

Reply:
> we can exact the connect operation into a function, instead of repeating it.

It's trivial and unnecessary to do this. Only connect and handshake functions are called twice on demand, and we need to control the error response comes from the second try.

Reply:
I see.